### ISSUE TITLE
make CompilationOptions and ParseOptions to implement IObjectWritable like diagnostics and syntax node

### ISSUE THREAD

**Issue by heejaechang** [dotnet/roslyn#12795]
having serializer for compiler objects in IDE side make it hard to maintain code. it would be nice if compiler directly support serialization.

**Comment by jaredpar** [comment]
@heejaechang is there a specific motivating scenario here or is it just a "nice to have"?

**Comment by heejaechang** [comment]
@jaredpar currently the serialization is done in IDE side which means if new flag is added to the compilation options, then IDE won't know about it. also, if flag added is not easily serializable, then it becomes issue for IDE.

having it in the compiler (closer to the source) make 1. harder to miss new flag, 2. people who adding the flag think about serialization.

imagine syntax node serialization is done in IDE side rather than compiler side, it would have been nightmare for IDE to make the serialization working properly all the time.

**Comment by heejaechang** [comment]
@jaredpar can it be done at some point? I can do what I have done for IDE (http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Execution/CSharpOptionsSerializationService.cs,19) if you want but I want some help from compiler team for some fields such as below.

// REVIEW: What should I do with these. we probably need to implement either out own one
            //         or somehow share these as service....
            //
            // XmlReferenceResolver xmlReferenceResolver
            // SourceReferenceResolver sourceReferenceResolver
            // MetadataReferenceResolver metadataReferenceResolver
            // AssemblyIdentityComparer assemblyIdentityComparer
            // StrongNameProvider strongNameProvider

the problem of them is that these are not data, but behavior (code) which I am not sure how I should serialize/deserialize.

I can do what I did for ProjectInfo where I separated data (ProjectInfo.Attributes) and others and only serialize data and not behavior.

but otherwise, it would be nice if compiler team can decide how they want to make the options (both compilationoptions and parseoptions) serializable.
